nvflare/fed_job.py

```diff
@@ -0,0 +1,274 @@
+
+    def _deploy(self, app: FedApp, target: str):
+        if not isinstance(app, FedApp):
+            raise ValueError(f"App needs to be of type `FedApp` but was type {type(app)}")
+
+        client_server_config = app.get_app_config()
+        if isinstance(client_server_config, ClientAppConfig):
+            app_config = FedAppConfig(server_app=None, client_app=client_server_config)
+            app_name = f"app_{target}"
+        elif isinstance(client_server_config, ServerAppConfig):
+            app_config = FedAppConfig(server_app=client_server_config, client_app=None)
+            app_name = "app_server"
+        else:
+            raise ValueError(
+                f"App needs to be of type `ClientAppConfig` or `ServerAppConfig` but was type {type(client_server_config)}"
+            )
+
+        self.job.add_fed_app(app_name, app_config)
+        self.job.set_site_app(target, app_name)
+
+    def _run_deploy(self):
+        if not self._deployed:
+            for target in self._deploy_map:
+                self._deploy(self._deploy_map[target], target)
+
+            self._deployed = True
+
+    def export_job(self, job_root):
+        self._run_deploy()
+        self.job.generate_job_config(job_root)
+
+    def simulator_run(self, workspace, threads: int = None):
+        self._run_deploy()
+
+        n_clients = len(self.clients)
+        if threads is None:
+            threads = n_clients
+
+        self.job.simulator_run(
+            workspace,
+            clients=",".join(self.clients),
+            n_clients=n_clients,
+            threads=threads,
+            gpu=",".join([self._gpus[client] for client in self._gpus.keys()]),
+        )
+
+
+class ExecutorApp(FedApp):
+    def __init__(self, external_scripts: List = None):
+        """Wrapper around `ClientAppConfig`.
+
+        Args:
+            external_scripts: List of external scripts that need to be deployed to the client. Defaults to None.
+        """
+        super().__init__()
+        self.external_scripts = external_scripts
+        self._create_client_app()
+
+    def add_executor(self, executor, tasks=None):
+        if tasks is None:
+            tasks = ["*"]  # Add executor for any task by default
+        self.app.add_executor(tasks, executor)
+
+    def _create_client_app(self):
+        self.app = ClientAppConfig()
+
+        component = FilePipe(  # TODO: support CellPipe, causes type error for passing secure_mode = "{SECURE_MODE}"
+            mode=Mode.PASSIVE,
+            root_path="{WORKSPACE}/{JOB_ID}/{SITE_NAME}",
+        )
+        self.app.add_component("metrics_pipe", component)
+
+        component = MetricRelay(pipe_id="metrics_pipe", event_type="fed.analytix_log_stats", read_interval=0.1)
+        self.app.add_component("metric_relay", component)
+
+        component = ExternalConfigurator(component_ids=["metric_relay"])
+        self.app.add_component("config_preparer", component)
```

Comment:
This will make the ExecutorApp limit to FilePipe, MetricRelay components only. Do we want to make this more generic?

Reply:
We want to change it to use CellPipe later by default. Maybe add an option to disable.

Reply:
With InProcessExecutor, Cellpipe is not needed.